### DISCUSSION TITLE
Process event when needed

### DISCUSSION THREAD

**Discussion by xland** [mikke89/RmlUi#417] [Q&A] [answered]
Should I block the following infinite loop  when the application does not need to process any event or animation?
```c++
bool running = true;
while (running)
{
      // Handle input and window events.
      running = Backend::ProcessEvents(context, &Shell::ProcessKeyDownShortcuts);
      //......
}
```
I just want to reduce CPU cust.
Could you give me some suggestions or tips?
Thanks very much!

SFML use `waitEvent` method to wait for an event
`waitEvent` function is blocking, if there's no pending event then it will wait until an event is received.
I don't know if RmlUi has a similar method, or whether it is good or not
```c++
 sf::Event event;
  while (window.waitEvent(event))   //waitEvent 
  {
      if (event.type == sf::Event::Closed)
          window.close();
      window.clear();
      window.draw(shape);
      window.display();
  }
```

**Comment by mikke89** [comment, marked as answer]
Yeah, I would suggest that you try it out and see how it goes. You can simply test it by modifying the SFML backend, it would just be a matter of swapping out the `pollEvent` with `waitEvent` here: https://github.com/mikke89/RmlUi/blob/1b1a8823f9a21b3966343bb760d6de8037f31ad9/Backends/RmlUi_Backend_SFML_GL2.cpp#L201
You could implement the same functionality in any of the other backends too. That will be platform specific, so you will have to look up how to do it for your platform (or see how SFML implements it).

Now, this clearly wouldn't work for example if you have animations playing. Then you would have to run the loop regardless of events. There could be other situations like this, like when you manually want to update the DOM.

**Reply by xland** [reply]
OK, Thanks.
I am using ` GLFW_ GL3 ` backend, I may try `SFML` backend in the future.

When I move the mouse over an element, the element's hover style is triggered.
At this time, I can't refresh the interface without modifying `RmlUi`.
It would be great if `RmlUi` supports this feature internally.

Chrome uses skia as the rendering engine.
When a web page has animation or a hover event is triggered (and some thing like these), it will consume CPU resources.
When the page does not contain these things, it will not consume CPU resources.

I don't know whether this feature is good or bad for RmlUi.

**Reply by mikke89** [reply]
It would certainly be a nice feature to have, I would be fully on board with it. It would take someone willing to spend that time to implement this in the library.

**Reply by xland** [reply]
Glad to hear you agree with this idea. 
Neither ImGui nor raylib has this feature , while Qt and FLTK have this feature.

I don't think I can handle this job. 
Hope other people can participate in it.

**Comment by mikke89** [comment]
I think you'll be happy to see that this feature is now integrated into RmlUi, through the pull request #436 (thanks @Thalhammer). 🥳  This will be included in the next release of RmlUi.

For example, on the SDL platform, you can now poll for events like shown here:
https://github.com/mikke89/RmlUi/blob/801b23945d36e7368c8f3df4653bee1b513c71d5/Backends/RmlUi_Backend_SDL_GL2.cpp#L266

We couldn't integrate this into the SFML platform, since the library seems to lack a timeout version of waitEvent, but please check out the other platforms. For usage on other platforms, check the pull request or backends source code. Documentation is being updated.

**Reply by xland** [reply]
Glad to hear that, but I've been working on other projects. 
I'll test it out when I have time.
